libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs

```diff
@@ -271,539 +398,578 @@ toUMapPair (UMap !m1 !m2) =
   , "pointers" .= m2
   ]
 
--- | It is worthwhile stating the invariant that holds on a Unified Map
---   The 'ptrmap' and the 'ptrT' field of the 'tripmap' are inverses.
+instance Crypto c => EncCBOR (UMap c) where
+  encCBOR UMap {..} =
+    encodeListLen 2 <> encodeMap encCBOR encCBOR umElem <> encodeMap encCBOR encCBOR umPtr
+
+instance Crypto c => DecShareCBOR (UMap c) where
+  type Share (UMap c) = (Interns (Credential 'Staking c), Interns (KeyHash 'StakePool c))
+  decSharePlusCBOR =
+    StateT
+      ( \(a, b) ->
+          decodeRecordNamed "UMap" (const 2) $ do
+            umElem <- decodeMap (interns a <$> decCBOR) (decShareCBOR b)
+            let a' = internsFromMap umElem <> a
+            umPtr <- decodeMap decCBOR (interns a' <$> decCBOR)
+            pure (UMap {..}, (a', b))
+      )
+
+-- | It is worthwhile stating the invariant that holds on a Unified Map.
+-- The `umPtr` and the `ptrT` field of the `umElem` are inverses.
 umInvariant :: Credential 'Staking c -> Ptr -> UMap c -> Bool
-umInvariant stake ptr (UMap tripmap ptrmap) = forwards && backwards
+umInvariant cred ptr UMap {..} = forwards && backwards
   where
     forwards =
-      case Map.lookup stake tripmap of
-        Nothing -> all (stake /=) ptrmap
-        Just (Triple _c set _d) ->
-          if Set.member ptr set
-            then case Map.lookup ptr ptrmap of
-              Nothing -> False
-              Just stake2 -> stake == stake2
-            else True
+      case Map.lookup cred umElem of
+        Nothing -> cred `notElem` umPtr
+        Just (UElemP _r ptrSet _s _) ->
+          not (Set.member ptr ptrSet)
+            || ( case Map.lookup ptr umPtr of
+                  Nothing -> False
+                  Just cred2 -> cred == cred2
+               )
     backwards =
-      case Map.lookup ptr ptrmap of
-        Nothing -> all (\(Triple _ set _) -> Set.notMember ptr set) tripmap
-        Just cred ->
-          case Map.lookup cred tripmap of
+      case Map.lookup ptr umPtr of
+        Nothing -> all (\(UElemP _ ptrSet _ _) -> Set.notMember ptr ptrSet) umElem
+        Just cred' ->
+          case Map.lookup cred' umElem of
             Nothing -> False
-            Just (Triple _ set _) -> Set.member ptr set
-
--- =====================================================
+            Just (UElemP _ ptrSet _ _) -> Set.member ptr ptrSet
 
--- VIEW
+-- $ UView
```

Comment:
Why the `$`?

It has a specific meaning of named section in haddock, but it doesn't look to be used like that.

Reply:
Noted! I have removed them now. I presume they make little sense if there are no accompanies notes to the placeholders.